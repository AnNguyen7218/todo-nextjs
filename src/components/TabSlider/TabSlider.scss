@import 'config.gutters';
@import 'config.colors';

.tab-headers-container {
  display: flex;
  cursor: pointer;
}

.header {
  display: flex;
  align-items: center;
  justify-content: center;
  border-bottom: 2px solid transparent;
  transition: all 0.4s;
  opacity: 0.24;

  &:not(:first-child) {
    margin-left: gutter(20);
  }

  &--selected {
    opacity: 1;
    transition: all 0.7s;
    border-color: $purple;
  }
}

.icon {
  padding: gutter(8) gutter(12);
  width: 3rem;
  height: 3rem;
}
