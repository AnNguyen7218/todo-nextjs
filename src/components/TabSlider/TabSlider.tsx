/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useState, useEffect } from 'react';
import cx from 'classnames';
import {
  ReactIdSwiperProps,
  ReactIdSwiperChildren,
  GetSwiper,
  SwiperInstance,
} from 'react-id-swiper';
import { Slider, Container } from 'components';
import css from './TabSlider.scss';

type Props = {
  tabList: TabSliderType[];
  children: ReactIdSwiperChildren;
  swiperParams?: ReactIdSwiperProps;
  getSwiper?: GetSwiper;
  wrapperClassName?: string;
  sliderContainerClassName?: string;
  getSelectedTag?: (index: number) => void;
  initialTab?: number;
};

const TabSlider: React.FC<Props> = ({
  tabList,
  children,
  swiperParams = {},
  wrapperClassName,
  sliderContainerClassName,
  getSelectedTag,
  initialTab = 0,
}) => {
  const [swiper, setSwiper] = useState<SwiperInstance>(null);
  const [selectedTabIndex, setSelectedTabIndex] = useState(initialTab);

  useEffect(() => {
    setSelectedTabIndex(swiper?.params.initialSlide || initialTab);
  }, [swiper]);

  useEffect(() => {
    getSelectedTag && getSelectedTag(selectedTabIndex);
  }, [selectedTabIndex]);

  const defaultSwiper: ReactIdSwiperProps = {
    loop: false,
    on: {
      slideChange: function() {
        const swiper: any = this;
        const currentIndex = swiper.realIndex;
        setSelectedTabIndex(currentIndex);
      },
    },
    initialSlide: initialTab,
    ...swiperParams,
  };

  const onClick = (index: number) => {
    swiper?.slideTo(index);
    setSelectedTabIndex(index);
  };

  return (
    <div className={cx(wrapperClassName)}>
      <Container className={css['tab-headers-container']}>
        {tabList.map((tab, index) => {
          const { icon: Icon } = tab;
          const isSelected = selectedTabIndex === index;
          return (
            <div
              key={index}
              className={cx(css['header'], {
                [css['header--selected']]: isSelected,
              })}
              onClick={() => onClick(index)}
              data-testid={`tab-${index}`}
            >
              <Icon className={css['icon']} />
            </div>
          );
        })}
      </Container>
      <div className={cx(sliderContainerClassName)}>
        <Slider swiperParams={defaultSwiper} getSwiper={setSwiper}>
          {children}
        </Slider>
      </div>
    </div>
  );
};

export default TabSlider;
