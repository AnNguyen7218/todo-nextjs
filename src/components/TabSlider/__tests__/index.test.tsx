import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import StartUpIcon from 'assets/icons/ic-startup.svg';
import ProfessionalIcon from 'assets/icons/ic-professional.svg';
import StudioIcon from 'assets/icons/ic-studio.svg';
import GrowthIcon from 'assets/icons/ic-rocket.svg';
import TabSlider from '../TabSlider';

afterEach(cleanup);

const headers = [
  { icon: StartUpIcon },
  { icon: ProfessionalIcon },
  { icon: StudioIcon },
  { icon: GrowthIcon },
];

describe('TabSlider component', () => {
  test('it should match snapshot', () => {
    const { container } = render(
      <TabSlider tabList={headers}>
        <p>Item1</p>
        <p>Item2</p>
        <p>Item3</p>
      </TabSlider>
    );
    expect(container).toMatchSnapshot();
  });

  test('it should match snapshot when set init tab', () => {
    const { container } = render(
      <TabSlider tabList={headers} swiperParams={{ initialSlide: 1 }}>
        <p>Item1</p>
        <p>Item2</p>
        <p>Item3</p>
      </TabSlider>
    );
    expect(container).toMatchSnapshot();
  });

  test('it should match snapshot when click on tab header', () => {
    const { asFragment, getByTestId } = render(
      <TabSlider tabList={headers}>
        <p>Item1</p>
        <p>Item2</p>
        <p>Item3</p>
      </TabSlider>
    );
    expect(asFragment()).toMatchSnapshot();
    const tab = getByTestId('tab-1');
    fireEvent.click(tab);
    expect(asFragment()).toMatchSnapshot();
  });
});
