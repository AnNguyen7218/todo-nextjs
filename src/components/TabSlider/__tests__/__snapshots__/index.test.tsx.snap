// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabSlider component it should match snapshot 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="container tab-headers-container"
    >
      <div
        class="header header--selected"
        data-testid="tab-0"
      >
        <ic-startup.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-1"
      >
        <ic-professional.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-2"
      >
        <ic-studio.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-3"
      >
        <ic-rocket.svg
          classname="icon"
        />
      </div>
    </div>
    <div
      class=""
    >
      <div
        class="swiper-container swiper-container-initialized swiper-container-horizontal"
      >
        <div
          class="swiper-wrapper"
          style="webkit-transition-duration: 0ms; transition-duration: 0ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
        >
          <p
            class="swiper-slide"
          >
            Item1
          </p>
          <p
            class="swiper-slide"
          >
            Item2
          </p>
          <p
            class="swiper-slide"
          >
            Item3
          </p>
        </div>
        <span
          aria-atomic="true"
          aria-live="assertive"
          class="swiper-notification"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TabSlider component it should match snapshot when click on tab header 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="container tab-headers-container"
    >
      <div
        class="header header--selected"
        data-testid="tab-0"
      >
        <ic-startup.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-1"
      >
        <ic-professional.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-2"
      >
        <ic-studio.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-3"
      >
        <ic-rocket.svg
          classname="icon"
        />
      </div>
    </div>
    <div
      class=""
    >
      <div
        class="swiper-container swiper-container-initialized swiper-container-horizontal"
      >
        <div
          class="swiper-wrapper"
          style="webkit-transition-duration: 0ms; transition-duration: 0ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
        >
          <p
            class="swiper-slide"
          >
            Item1
          </p>
          <p
            class="swiper-slide"
          >
            Item2
          </p>
          <p
            class="swiper-slide"
          >
            Item3
          </p>
        </div>
        <span
          aria-atomic="true"
          aria-live="assertive"
          class="swiper-notification"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TabSlider component it should match snapshot when click on tab header 2`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="container tab-headers-container"
    >
      <div
        class="header"
        data-testid="tab-0"
      >
        <ic-startup.svg
          classname="icon"
        />
      </div>
      <div
        class="header header--selected"
        data-testid="tab-1"
      >
        <ic-professional.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-2"
      >
        <ic-studio.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-3"
      >
        <ic-rocket.svg
          classname="icon"
        />
      </div>
    </div>
    <div
      class=""
    >
      <div
        class="swiper-container swiper-container-initialized swiper-container-horizontal"
      >
        <div
          class="swiper-wrapper"
          style="webkit-transition-duration: 300ms; transition-duration: 300ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
        >
          <p
            class="swiper-slide"
          >
            Item1
          </p>
          <p
            class="swiper-slide"
          >
            Item2
          </p>
          <p
            class="swiper-slide"
          >
            Item3
          </p>
        </div>
        <span
          aria-atomic="true"
          aria-live="assertive"
          class="swiper-notification"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TabSlider component it should match snapshot when set init tab 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="container tab-headers-container"
    >
      <div
        class="header"
        data-testid="tab-0"
      >
        <ic-startup.svg
          classname="icon"
        />
      </div>
      <div
        class="header header--selected"
        data-testid="tab-1"
      >
        <ic-professional.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-2"
      >
        <ic-studio.svg
          classname="icon"
        />
      </div>
      <div
        class="header"
        data-testid="tab-3"
      >
        <ic-rocket.svg
          classname="icon"
        />
      </div>
    </div>
    <div
      class=""
    >
      <div
        class="swiper-container swiper-container-initialized swiper-container-horizontal"
      >
        <div
          class="swiper-wrapper"
          style="webkit-transition-duration: 0ms; transition-duration: 0ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
        >
          <p
            class="swiper-slide"
          >
            Item1
          </p>
          <p
            class="swiper-slide"
          >
            Item2
          </p>
          <p
            class="swiper-slide"
          >
            Item3
          </p>
        </div>
        <span
          aria-atomic="true"
          aria-live="assertive"
          class="swiper-notification"
        />
      </div>
    </div>
  </div>
</div>
`;
